src/AlgebraicGeometry/Surfaces/EnriquesAut.jl

```diff
@@ -0,0 +1,908 @@
+  Rtilde = primitive_closure(R, lattice(V,basis_matrix(Sm)))
+  return R, root_lattice_recognition(R), root_lattice_recognition(Rtilde)
+end
+
+@doc raw"""
+    mass(ECtx::EnriquesBorcherdsCtx)
+    
+Return the mass of ``Y`` 
+
+Let ``V_0`` be a complete set of representatives of the ``\mathrm{Aut}(Y)``-orbits of ``L_{26}|S_Y``-chambers.
+The mass satisfies
+```math
+\mathrm{mass}(Y)=\sum_{D \in V_0} \frac{1}{\sharp \mathrm{Aut}_G(D)} = \frac{\mathrm{ind}(D_0)}{\sharp \bar G_{X-}}
+```
+where ``D_0`` is the initial chamber and ``\mathrm{inv}(D_0)`` is its volume index. 
+"""
+function mass(ECtx::EnriquesBorcherdsCtx)
+  return chamber_invariants(ECtx)[1]//ECtx.orderGbar
+end
+
+function membership_test_as_group(data::EnriquesBorcherdsCtx, f::FqMatrix)
+  # test if f lies in Gplus
+  # f must
+  # act as identity on Dplus^perp
+  b =  f in data.Gplus_mat
+  return b
+end
+
+function initial_rays(ECtx::EnriquesBorcherdsCtx)
+  if !isdefined(ECtx, :initial_rays)
+    ECtx.initial_rays = rays(ECtx.initial_chamber)
+  end
+  return ECtx.initial_rays
+end
+
+function initial_isotropic_rays(ECtx::EnriquesBorcherdsCtx)
+  if !isdefined(ECtx, :initial_isotropic_rays)
+    iso_rays = filter(r->0==r*ECtx.gramSY*transpose(r), rays(ECtx.initial_chamber))
+    ECtx.initial_isotropic_rays = iso_rays
+  end
+  return ECtx.initial_isotropic_rays
+end
+
+
+function _assure_membership_test_as_set(ECtx::EnriquesBorcherdsCtx)
+  if isdefined(ECtx,:imgs_mod2)
+    return
+  end
+  @vprint :EnriquesAuto 2 "computing $(order(ECtx.Gplus_mat)) images mod 2\n"
+  imgs_mod2 = Set{FqMatrix}()
+  for g in ECtx.Gplus_mat
+    h = matrix(g)
+    push!(imgs_mod2, ECtx.Dplus*h)
+  end
+  ECtx.imgs_mod2 = imgs_mod2
+end
+
+function membership_test_set(data::EnriquesBorcherdsCtx, f::FqMatrix)
+  # test if f lies in Gplus
+  # f must
+  # act as identity on Dplus^perp
+  data.Dplus_perp*f == data.Dplus_perp || return false
+  # its restriction to Dplus
+  # must lie in in Gplus_res
+  f_of_Dplus = data.Dplus*f
+  return f_of_Dplus in data.imgs_mod2
+end
+
+
+
+# in principle it would be enough to just store
+# tau and parent wall
+# then tau can be recomputed via the spanning tree
+# that would reduce storage by a factor of about 10
+# ... but increase computation time
+
+@doc raw""" 
+    EnriquesChamber 
+
+An ``S_Y(2)|L_{26}`` chamber.
+
+The chamber is represented as ``D_0^\tau`` for ``D_0`` 
+the initial chamber and ``\tau \in O(S_Y)``. 
+"""
+struct EnriquesChamber  # no need to make this mutable.
+  data::EnriquesBorcherdsCtx
+  tau::ZZMatrix
+  parent_wall::ZZMatrix # for the spanning tree, the wall via which this chamber was computed
+end
+
+function ==(x::EnriquesChamber, y::EnriquesChamber)
+  @req x.data === y.data "Chambers must be associated to the same Enriques surface"
+  return x.tau == y.tau
+end
+
+function Base.hash(D::EnriquesChamber, h::UInt)
+  return hash(fingerprint(D), h)
```

Comment:
```suggestion
  return hash((D.data, D.tau), h)
```
If `==` requires that `.tau` is equal, you can just use that in the hash as well. This makes `hash` a lot faster to compute, since it avoids `fingerprint`. You even note that `fingerprint` is "a bit expensive" to compute, but `hash` should be fast.

Reply:
true